src/lib/png/png.c

```diff
@@ -0,0 +1,1129 @@
+		bitlen[i] = 9;
+	}
+
+	for(i = 256; i <= 279; i++) {
+		bitlen[i] = 7;
+	}
+
+	for(i = 280; i <= 287; i++) {
+		bitlen[i] = 8;
+	}
+
+	err = make_tree_first_step(tree, bitlen, NUMBER_OF_CODE_SYMBOLS, 15);
+	return err;
+}
+
+/* get the trees with fixed huffman blocks */
+int build_fixed_trees(HuffmanTree *ll_tree, HuffmanTree *dist_tree) {
+	int err;
+	err = build_fixed_ll_tree(ll_tree);
+	if (!err) {
+		err = build_fixed_dist_tree(dist_tree);
+	}
+	return err;
+}
+
+int inflate_huffman(unsigned char *out, unsigned char *in, int *bp,
+								int *pos, int in_length, int btype) {
+	int err = 0;
+	HuffmanTree ll_tree; /* literal and length codes */
+	HuffmanTree dist_tree; /* distance codes */
+	int in_bitlen = in_length * 8;
+	int l_code, length_base; /* ll code and length */
+	int dist_code, dist; /* distance code and distance */
+	int addit_bits_ll, addit_bits_d; /* extra bits for length and distance */
+	int first_pos, forw, back;
+
+	/* build Huffman trees */
+	HuffmanTree_init(&ll_tree);
+	HuffmanTree_init(&dist_tree);
+
+	if(btype == 1) {
+		err = build_fixed_trees(&ll_tree, &dist_tree);
+	} else if(btype == 2) {
+		err = build_dynamic_trees(&ll_tree, &dist_tree, in, bp, in_length);
+	}
+
+	while(true) { /* decode symbols until error or end */
+		l_code = huffman_decode_symbol(in, bp, &ll_tree, in_bitlen);  /* lit symbol */
+
+		if(l_code <= 255) {
+			out[*pos] = (unsigned char)l_code;
+			(*pos)++;
+		} else if(l_code >= FIRST_LENGTH_INDEX && l_code <= LAST_LENGTH_INDEX) {
+
+			/* step 1: get length_base */
+			length_base = LENGTH_BASE[l_code - FIRST_LENGTH_INDEX];
+
+			/* step 2: get additional bits and add them value to length_base */
+			addit_bits_ll = LENGTH_ADDITIONAL[l_code - FIRST_LENGTH_INDEX];
+			length_base += read_bits_from_stream(bp, in, addit_bits_ll);
+
+			/* step 3: get distance code and distance */
+			dist_code = huffman_decode_symbol(in, bp, &dist_tree, in_bitlen);
+
+			if(dist_code == -1) { /* decode_symbol return -1 */
+				err = 2; /* TODO table of errors */
+				break;
+			}
+
+			if(dist_code > 29) { /* distance code cant be 30-32 */
+				err = 3;
+				break;
+			}
+			dist = DISTANCE_BASE[dist_code];
+
+			/* step 4: get extra bits from distance */
+			addit_bits_d = DISTANCE_ADDITIONAL[dist_code];
+			dist += read_bits_from_stream(bp, in, addit_bits_d);
+
+			/* step 5: fill in all the out[n] values based on the length and dist */
+			first_pos = (*pos);
+			back = first_pos - dist;
+			if (dist < length_base) {
+				for(forw = 0; forw < length_base; forw++) {
+					out[(*pos)++] = out[back++];
+				}
+			} else {
+				memcpy(out + *pos, out + back, length_base);
+				*pos += length_base;
+			}
+
+		} else if (l_code == 256) {
+			break; /* end of code */
+
+		} else if (l_code == -1) { /* decodesymbol return -1 */
+			err = 2;
+			break;
+		}
+	}
+	HuffmanTree_cleanup(&ll_tree);
+	HuffmanTree_cleanup(&dist_tree);
+
+	return err;
+}
+
+unsigned char predictor(short a, short b, short c) {
```

Comment:
I guess arguments may be declared as `uint8_t`, so there will be no need in cast in return values.

Reply:
replace all unsigned to uint